# Checkpoint Three: Cleaning Data

Now you are ready to clean your data. Before starting coding, provide the link to your dataset below.

My dataset: https://www.kaggle.com/datasets/arbelaezch/ncaam-basketball-game-schedules-2014-2024

Import the necessary libraries and create your dataframe(s).

In [166]:
# Importing Packages
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Importing Datasets to eventually combine.
df13_14 = pd.read_csv("../final-project-data/2013-14_schedule.csv")
df14_15 = pd.read_csv("../final-project-data/2014-15_schedule.csv")
df15_16 = pd.read_csv("../final-project-data/2015-16_schedule.csv")
df16_17 = pd.read_csv("../final-project-data/2016-17_schedule.csv")
df17_18 = pd.read_csv("../final-project-data/2017-18_schedule.csv")
df18_19 = pd.read_csv("../final-project-data/2018-19_schedule.csv")
df20_21 = pd.read_csv("../final-project-data/2020-21_schedule.csv")
df21_22 = pd.read_csv("../final-project-data/2021-22_schedule.csv")
df22_23 = pd.read_csv("../final-project-data/2022-23_schedule.csv")
df23_24 = pd.read_csv("../final-project-data/2023-24_schedule.csv")

# Adding New Columns
df13_14["Season"] = "2013-14"
df14_15["Season"] = "2014-15"
df15_16["Season"] = "2015-16"
df16_17["Season"] = "2016-17"
df17_18["Season"] = "2017-18"
df18_19["Season"] = "2018-19"
df20_21["Season"] = "2020-21"
df21_22["Season"] = "2021-22"
df22_23["Season"] = "2022-23"
df23_24["Season"] = "2023-24"

# Putting together list of dataframes to apply items to later...
dataframes_list = [df13_14, df14_15, df15_16, df16_17, df17_18, df18_19, df20_21, df21_22, df22_23, df23_24]

# Combining into mega-dataframe to work with; this will be updated later.
df_mega = pd.concat(dataframes_list, ignore_index= True)

In [167]:
print(df_mega.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108110 entries, 0 to 108109
Data columns (total 26 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Date                108110 non-null  object 
 1   Type                108110 non-null  object 
 2   Team                108110 non-null  object 
 3   Conf.               108110 non-null  object 
 4   Opp.                108110 non-null  object 
 5   Venue               108110 non-null  object 
 6   Adj. O              108110 non-null  float64
 7   Adj. D              108110 non-null  float64
 8   T                   108110 non-null  float64
 9   O.EFF               108110 non-null  float64
 10  O.eFG%              108110 non-null  float64
 11  O.TO%               108110 non-null  float64
 12  O.Reb%              108110 non-null  float64
 13  O.FTR               108110 non-null  float64
 14  D.EFF               108110 non-null  float64
 15  D.eFG%              108110 non-nul

## Missing Data

Test your dataset for missing data and handle it as needed. Make notes in the form of code comments as to your thought process.

In [168]:
# Looking for missing data - I do not believe we have any, but we'll check that below. 
for col in df_mega.columns:
    total_null = np.sum(df_mega[col].isnull())
    print('{} - {}'.format(col, total_null))

Date - 0
Type - 0
Team - 0
Conf. - 0
Opp. - 0
Venue - 0
Adj. O - 0
Adj. D - 0
T - 0
O.EFF - 0
O.eFG% - 0
O.TO% - 0
O.Reb% - 0
O.FTR - 0
D.EFF - 0
D.eFG% - 0
D.TO% - 0
D.Reb% - 0
D.FTR - 0
G-Sc - 0
+/- - 0
Win - 0
Points_For - 0
Points_Against - 0
Point_Differential - 0
Season - 0


## Irregular Data

Detect outliers in your dataset and handle them as needed. Use code comments to make notes about your thought process.

In [169]:
# looking at the output to see if there's anything that stands out as an outlier
df_mega.describe()

Adj. O         Adj. D              T          O.EFF  \
count  108110.000000  108110.000000  108110.000000  108110.000000   
mean      103.839220     103.095438      68.192722     103.379821   
std        13.982832      13.478265       5.671272      15.303539   
min        47.000000      46.100000      46.700000      34.900000   
25%        94.400000      94.000000      64.300000      93.200000   
50%       103.600000     103.000000      68.000000     103.500000   
75%       113.000000     112.100000      71.800000     113.600000   
max       167.200000     166.100000      97.700000     174.200000   

              O.eFG%          O.TO%         O.Reb%          O.FTR  \
count  108110.000000  108110.000000  108110.000000  108110.000000   
mean       50.356156      18.300381      28.943264      35.109465   
std         8.656261       5.445724       9.228174      15.689865   
min        12.200000       0.000000       0.000000       0.000000   
25%        44.400000      14.500000      22.600000      23.800000   
50%        50.000000      18.000000      28.600000      33.300000   
75%        56.100000      21.800000      35.100000      44.100000   
max        91.700000      49.600000      69.100000     184.800000   

               D.EFF         D.eFG%          D.TO%         D.Reb%  \
count  108110.000000  108110.000000  108110.000000  108110.000000   
mean      103.379821      50.356156      18.300381      28.943264   
std        15.303539       8.656261       5.445724       9.228174   
min        34.900000      12.200000       0.000000       0.000000   
25%        93.200000      44.400000      14.500000      22.600000   
50%       103.500000      50.000000      18.000000      28.600000   
75%       113.600000      56.100000      21.800000      35.100000   
max       174.200000      91.700000      49.600000      69.100000   

               D.FTR           G-Sc            Win     Points_For  \
count  108110.000000  108110.000000  108110.000000  108110.000000   
mean       35.109465      50.931547       0.500000      77.084488   
std        15.689865      31.640339       0.500002      10.798666   
min         0.000000       0.000000       0.000000      38.000000   
25%        23.800000      21.400000       0.000000      70.000000   
50%        33.300000      51.200000       0.500000      77.000000   
75%        44.100000      80.700000       1.000000      84.000000   
max       184.800000     100.000000       1.000000     149.000000   

       Points_Against  Point_Differential  
count   108110.000000       108110.000000  
mean        65.197725           11.886763  
std         10.745424            9.153406  
min         24.000000            1.000000  
25%         58.000000            5.000000  
50%         65.000000           10.000000  
75%         72.000000           16.000000  
max        144.000000           80.000000

The only items that are relevant for what I'm looking into are Win, Points For, Points Against, and Point Differential. I'm going to drop a few of these other columns to decrease the noise in the next section, but for now the biggest issues I see are that Wins should be a Boolean value. The Date should also be updated to be a date, and not a general object.

In [170]:
df_mega['Win']= df_mega['Win'].astype('bool')
df_mega['Date'] = pd.to_datetime(df_mega['Date'])
df_mega.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108110 entries, 0 to 108109
Data columns (total 26 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   Date                108110 non-null  datetime64[ns]
 1   Type                108110 non-null  object        
 2   Team                108110 non-null  object        
 3   Conf.               108110 non-null  object        
 4   Opp.                108110 non-null  object        
 5   Venue               108110 non-null  object        
 6   Adj. O              108110 non-null  float64       
 7   Adj. D              108110 non-null  float64       
 8   T                   108110 non-null  float64       
 9   O.EFF               108110 non-null  float64       
 10  O.eFG%              108110 non-null  float64       
 11  O.TO%               108110 non-null  float64       
 12  O.Reb%              108110 non-null  float64       
 13  O.FTR               108110 no

## Unnecessary Data

Look for the different types of unnecessary data in your dataset and address it as needed. Make sure to use code comments to illustrate your thought process.

Here's where things get interesting... we're going to first remove all the columns we don't need. As a reminder, the different columns are listed below.

In [171]:
list(df_mega.columns)
print(df_mega.shape)
df_mega = df_mega.drop(columns = ['Adj. O',
 'Adj. D',
 'T',
 'O.EFF',
 'O.eFG%',
 'O.TO%',
 'O.Reb%',
 'O.FTR',
 'D.EFF',
 'D.eFG%',
 'D.TO%',
 'D.Reb%',
 'D.FTR',
 'G-Sc',
 '+/-'])
df_mega.head(5)

(108110, 26)


Date Type          Team Conf.            Opp.    Venue   Win  \
0 2013-11-08    N          Duke   ACC        Davidson     Home  True   
1 2013-11-08    N     Wisconsin   B10      St. John's  Neutral  True   
2 2013-11-08    N  Saint Mary's   WCC  Louisiana Tech     Home  True   
3 2013-11-08    N    Fresno St.   MWC       UC Irvine     Away  True   
4 2013-11-08    N   Connecticut  Amer        Maryland  Neutral  True   

   Points_For  Points_Against  Point_Differential   Season  
0         111              77                  34  2013-14  
1          86              75                  11  2013-14  
2          83              70                  13  2013-14  
3          98              97                   1  2013-14  
4          78              77                   1  2013-14

Additionally, during EDA I discovered that there are many games that are not relevant for my analysis. These are the conference and conference tournament games. I will now go ahead and remove those datapoints as well.

In [172]:
# Seeing what the different types of 'Type', or game type, there are.
print(df_mega.Type.unique())
print(df_mega.Venue.unique())
# We are removing the 'C' and 'CT' values, as these stand for conference and conference tournament.
filter_values = ['N','T','P']
df_mega = df_mega[df_mega['Type'].isin(filter_values)]
print(df_mega.shape)
df_mega.info

['N' 'C' 'CT' 'P' 'T']
['Home' 'Neutral' 'Away']
(40608, 11)


<bound method DataFrame.info of              Date Type          Team Conf.            Opp.    Venue    Win  \
0      2013-11-08    N          Duke   ACC        Davidson     Home   True   
1      2013-11-08    N     Wisconsin   B10      St. John's  Neutral   True   
2      2013-11-08    N  Saint Mary's   WCC  Louisiana Tech     Home   True   
3      2013-11-08    N    Fresno St.   MWC       UC Irvine     Away   True   
4      2013-11-08    N   Connecticut  Amer        Maryland  Neutral   True   
...           ...  ...           ...   ...             ...      ...    ...   
108105 2024-04-06    T       Alabama   SEC     Connecticut  Neutral  False   
108106 2024-04-06    T        Purdue   B10      N.C. State  Neutral   True   
108107 2024-04-06    T    N.C. State   ACC          Purdue  Neutral  False   
108108 2024-04-08    T   Connecticut    BE          Purdue  Neutral   True   
108109 2024-04-08    T        Purdue   B10     Connecticut  Neutral  False   

        Points_For  Points_Agai

That's much easier to look at and work with! During my EDA portion of the final project, I ran into an issue where each game is in the database twice. Below I will try and rectify that issue. There are 40608 rows now, so that means that I should have 20304 rows once duplicates are removed below.

In [173]:
# Creating a unique game ID for each game... this should help eliminate the duplicates.
df_mega["game_id"] = (
    df_mega[["Team","Opp."]].apply(lambda x: "_".join(sorted(x)), axis = 1)
)

# Based on some research I did, the best way to get the opponent conference is to "self-merge" from the duplicate row. Code below.
df_mega = df_mega.merge(
    df_mega[["game_id", "Date", "Team", "Conf."]],
    left_on = ["game_id", "Date","Opp."],
    right_on = ["game_id", "Date", "Team"],
    how = "left",
    suffixes = ("","_opp")
)

df_mega = df_mega.rename(columns = {"Conf._opp":"Opp_Conf"})

# We're also going to preserve the neutral-site games, because we're ultimately going to remove the remainder of the away games
venue_rank = {
    "Home": 0,
    "Neutral" : 1,
    "Away": 2
}

df_mega['venue_rank'] = df_mega["Venue"].map(venue_rank)



In [174]:
# Finally, we'll drop the duplicate rows!
df_mega = df_mega.sort_values("venue_rank").drop_duplicates(subset=["game_id","Date"]).drop(columns=["venue_rank","game_id","Team_opp"])

In [175]:
df_mega.head(4)
df_mega.shape

(20304, 12)

Lo and behold we have the correct number of rows! Huzzah!

## Inconsistent Data

Check for inconsistent data and address any that arises. As always, use code comments to illustrate your thought process.

In [176]:
#let's now go back and look at the descriptive stats for inconsistent data...
df_mega.describe()

Date    Points_For  Points_Against  \
count                          20304  20304.000000    20304.000000   
mean   2018-09-02 16:28:47.659574528     78.245370       64.537037   
min              2013-11-08 00:00:00     40.000000       25.000000   
25%              2015-11-26 00:00:00     71.000000       57.000000   
50%              2017-12-09 00:00:00     78.000000       64.000000   
75%              2021-12-04 00:00:00     85.000000       72.000000   
max              2024-04-08 00:00:00    144.000000      120.000000   
std                              NaN     11.273135       10.961850   

       Point_Differential  
count        20304.000000  
mean            13.708333  
min              1.000000  
25%              6.000000  
50%             11.000000  
75%             19.000000  
max             80.000000  
std             10.500579

While this does allow for a large range of scores etc., none of the numbers stand out to me as extreme or inconsistent.

## Summarize Your Results

Make note of your answers to the following questions.

1. Did you find all four types of dirty data in your dataset?
    I did not find missing data, and I'm not sure I would say that it gave me inconsistent data... but I definitely ran into Unnecessary data and a bit of irregular data.
2. Did the process of cleaning your data give you new insights into your dataset?
    The process did give me insight in that now I have a much better understanding of data manipulation in python -- this is something that I definitely needed refreshing on so this project was very helpful.
3. Is there anything you would like to make note of when it comes to manipulating the data and making visualizations?
    I would love to now go ahead and do actual statistical analysis on the data, now that I have a clean dataset. This would include visualizations such as bar charts and large tables showing head-to-heads... I wonder about clustering and whether certain conferences should be grouped together based on the winning percentage/outcomes. There's so much more than can be done now - I will definitely be keeping this going on the side once I have more time :-)